internal/state/graph/httproute.go

```diff
@@ -402,33 +402,91 @@ func findAcceptedHostnames(listenerHostname *v1beta1.Hostname, routeHostnames []
 		return []string{hostname}
 	}
 
-	match := func(h v1beta1.Hostname) bool {
-		if hostname == "" {
-			return true
-		}
+	var result []string
 
+	for _, h := range routeHostnames {
 		routeHost := string(h)
-		return routeHost == hostname || wildcardMatch(hostname, routeHost) || wildcardMatch(routeHost, hostname)
+		if match(hostname, routeHost) {
+			result = append(result, GetMoreSpecificHostname(hostname, routeHost))
+		}
 	}
 
-	var result []string
+	return result
+}
 
-	for _, h := range routeHostnames {
-		if match(h) {
-			if len(hostname) > len(h) {
-				result = append(result, hostname)
-			} else {
-				result = append(result, string(h))
+func match(listenerHost, routeHost string) bool {
+	if listenerHost == "" {
+		return true
+	}
+
+	if routeHost == listenerHost {
+		return true
+	}
+
+	wildcardMatch := func(host1, host2 string) bool {
+		return strings.HasPrefix(host1, "*.") && strings.HasSuffix(host2, strings.TrimPrefix(host1, "*"))
+	}
+
+	// check if listenerHost is a wildcard and routeHost matches
+	if wildcardMatch(listenerHost, routeHost) {
+		return true
+	}
+
+	// check if routeHost is a wildcard and listener matchess
+	return wildcardMatch(routeHost, listenerHost)
+}
+
+// GetMoreSpecificHostname returns the more specific hostname between the two inputs.
+//
+// This function assumes that the two hostnames match each other, either:
+// - Exactly
+// - One as a substring of the other
+// - Both as substrings of some parent wildcard
+func GetMoreSpecificHostname(hostname1, hostname2 string) string {
+	if hostname1 == hostname2 {
+		return hostname1
+	}
+
+	if hostname1 == "" {
+		return hostname2
+	} else if hostname2 == "" {
+		return hostname1
+	}
+
+	// Compare if wildcards are present
+	if strings.HasPrefix(hostname1, "*.") {
+		if strings.HasPrefix(hostname2, "*.") {
+			subdomains1 := strings.Split(hostname1, ".")
+			subdomains2 := strings.Split(hostname2, ".")
+
+			// Compare number of subdomains
+			if len(subdomains1) > len(subdomains2) {
+				return hostname1
 			}
+
+			return hostname2
 		}
+
+		return hostname2
+	} else if strings.HasPrefix(hostname2, "*.") {
+		return hostname1
 	}
 
-	return result
-}
+	subdomains1 := strings.Split(hostname1, ".")
```

Comment:
when we get to line 475, neither of the hostnames are wildcard. They are also not equal. I wonder if this is a legit case here (not a programmer error).  If is not, shall we panic?

Reply:
This was originally the case handled when a route binds to multiple listeners, and we pick the most specific listener. However, I don't actually think we need to do that.

e.g. if our route hostname is `*.example.com` (and empty section name), and we have two listeners `somelongname.example.com` and `foo.bar.example.com`, my original thoughts were that we only bind to the second one. But actually we should be binding to both I think. So I can probably remove this.